aea/test_tools/test_cases.py

```diff
@@ -0,0 +1,261 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains test case classes based on pytest for AEA end-to-end testing."""
+
+import os
+import shutil
+import signal
+import subprocess  # nosec
+import sys
+import tempfile
+from pathlib import Path
+
+import pytest
+
+from aea.cli import cli
+from aea.cli.common import DEFAULT_REGISTRY_PATH
+from aea.configurations.base import PublicId
+from aea.mail.base import Envelope
+from aea.test_tools.exceptions import AEATestingException
+from aea.test_tools.generic import encode_envelope
+
+from tests.common.click_testing import CliRunner
+from tests.conftest import AUTHOR, CLI_LOG_OPTION
+
+
+class AEATestCase:
+    """Test case for AEA end-to-end tests."""
+
+    @classmethod
+    def setup_class(cls, packages_dir_path=DEFAULT_REGISTRY_PATH):
+        """Set up the test class."""
+        cls.runner = CliRunner()
+        cls.cwd = os.getcwd()
+        cls.t = tempfile.mkdtemp()
+
+        # add packages folder
+        packages_src = os.path.join(cls.cwd, packages_dir_path)
+        packages_dst = os.path.join(cls.t, packages_dir_path)
+        shutil.copytree(packages_src, packages_dst)
+
+        cls.subprocesses = []
+
+        os.chdir(cls.t)
+
+    @classmethod
+    def teardown_class(cls):
+        """Teardown the test."""
+        cls._terminate_subprocesses()
+
+        os.chdir(cls.cwd)
+        try:
+            shutil.rmtree(cls.t)
+        except (OSError, IOError):
+            pass
+
+    @classmethod
+    def _terminate_subprocesses(cls):
+        """Terminate all launched subprocesses."""
+        for process in cls.subprocesses:
+            process.send_signal(signal.SIGINT)
+            process.wait(timeout=20)
```

Comment:
These two lines (and the next) should not be here, we usually want to test for proper exit in the actual test.

Reply:
Agree.